src/plugins/kdb/ldap/libkdb_ldap/ldap_principal2.c

```diff
@@ -1375,64 +1493,221 @@ krb5_read_tkt_policy(krb5_context context, krb5_ldap_context *ldap_context,
-        }
-        noofkeys += n_kd;
-        tmp = key_data;
-        /* Allocate an extra key data to avoid allocating zero bytes. */
-        key_data = realloc(key_data, (noofkeys + 1) * sizeof (krb5_key_data));
-        if (key_data == NULL) {
-            key_data = tmp;
-            st = ENOMEM;
+        err = asn1_decode_sequence_of_keys(&in, &keysets[ki]);
+        if (err)
             goto cleanup;
-        }
-        for (j = 0; j < n_kd; j++)
-            key_data[noofkeys - n_kd + j] = kd[j];
-        free (kd);
+
+        if (total_keys_out)
+            total_keys += keysets[ki].n_key_data;
+        ki++;
     }
 
-    entries->n_key_data = noofkeys;
+    if (total_keys_out)
+        *total_keys_out = total_keys;
+
+    *n_keysets_out = n_keys;
+    *keysets_out = keysets;
+    keysets = NULL;
+    n_keys = 0;
+
+cleanup:
+    free_ldap_seqof_key_data(keysets, n_keys, err);
+    return err;
+}
+
+krb5_error_code
+krb5_decode_krbsecretkey(krb5_context context, krb5_db_entry *entries,
+                         struct berval **bvalues, krb5_kvno *mkvno)
+{
+    char *user = NULL;
+    krb5_key_data *key_data = NULL, *tmp;
+    krb5_error_code err = 0;
+    ldap_seqof_key_data *keysets = NULL;
+    krb5_int16 i, n_keysets = 0, total_keys;
+
+    err = krb5_unparse_name(context, entries->princ, &user);
+    if (err)
+        goto cleanup;
+
+    err = decode_keys(bvalues, &keysets, &n_keysets, &total_keys);
+    if (err != 0) {
+        k5_prependmsg(context, err, error_message(err));
+        k5_setmsg(context, err,
+                  _("unable to decode stored principal key data"));
+        goto cleanup;
+    }
+
+    key_data = k5calloc(total_keys, sizeof(krb5_key_data), &err);
+    if (key_data == NULL)
+        goto cleanup;
+    memset(key_data, 0, total_keys * sizeof(krb5_key_data));
+
+    if (n_keysets > 0)
+        *mkvno = keysets[0].mkvno;
+
+    tmp = key_data;
+    for (i = 0; i < n_keysets; i++) {
+        memcpy(tmp, keysets[i].key_data,
+               sizeof(krb5_key_data) * keysets[i].n_key_data);
+        tmp += keysets[i].n_key_data;
+    }
+    entries->n_key_data = total_keys;
     entries->key_data = key_data;
 
+    key_data = NULL;
+
 cleanup:
-    free (user);
-    return st;
+    free(user);
+    free_ldap_seqof_key_data(keysets, n_keysets, err);
+    /*
+     * If an error occured free_keysets should have freed all the key contents,
+     * so the key_data alone can be freed here.
+     */
+    free(key_data);
+    return err;
+}
+
+static int
+compare_descending_osa_pw_hist_ent(const void *left_in, const void *right_in)
+{
+    osa_pw_hist_ent *left = (osa_pw_hist_ent *) left_in;
+    osa_pw_hist_ent *right = (osa_pw_hist_ent *) right_in;
+
+    /*
+     * qsort sorts in ascending order, but we need descending. Subtracting the
+     * left from the right will give us the desired sorting order.
+     */
+    return right->key_data[0].key_data_kvno - left->key_data[0].key_data_kvno;
```

Comment:
I am nervous at best, about assuming that the osa_pw_hist_ent pointers are non-empty.  I don't think decode_keys() or krb5_decode_krbsecretkey() enforces this, so in theory it's possible to read in zero-length.

Treating any empty entries as having kvno 0 would be reasonable:

    kvno_left = left->n_key_data ? left->key_data[0].key_data_kvno[0] : 0;
    kvno_right = (same thing for right)
    return kvno_right - kvno_left;

Also, if you take the above, make sure to use "int" as the type of kvno_left and kvno_right.  Right now you're subtracting two uint16_t values, so you never produce a negative result.  I'm not sure why the tests are passing.